### ISSUE TITLE
Release annotation fails right after AppInsights are deployed

### ISSUE THREAD

**Issue by M0ns1gn0r** [microsoft/ApplicationInsights-Home#438]
I am deploying AppInsights via ARM, it takes about 25 seconds to do the deployment, so I am expecting it took that time to configure and initialize itself.

After about 2.5 minutes I Deploy a function app which uses this instance of AppInsights. This deployment seems to include the implicit "create release annotation" step as the following error gets logged:

```
##[debug][PUT]https://management.azure.com/subscriptions/.../resourceGroups/cp-test/providers/microsoft.insights/components/cp-test/Annotations?api-version=2015-05-01
##[debug]addReleaseAnnotation. Data : {"statusCode":404,"statusMessage":"Not Found","headers":{"cache-control":"no-cache","pragma":"no-cache","content-length":"153","content-type":"application/json; charset=utf-8","expires":"-1","x-content-type-options":"nosniff","request-context":"appId=cid-v1:920e14b1-13f3-461a-a4bb-b4fe6f1a4525","access-control-expose-headers":"Request-Context","strict-transport-security":"max-age=31536000; includeSubDomains","x-ms-ratelimit-remaining-subscription-writes":"1199","server":"Microsoft-IIS/10.0","x-ms-request-id":"4d06217a-318c-44d8-99d4-f9be9a897d5c","x-ms-correlation-request-id":"4d06217a-318c-44d8-99d4-f9be9a897d5c","x-ms-routing-request-id":"SOUTHCENTRALUS:20191209T153700Z:4d06217a-318c-44d8-99d4-f9be9a897d5c","date":"Mon, 09 Dec 2019 15:36:59 GMT","connection":"close"},"body":{"code":"NotFound","message":"Not Found","innererror":{"diagnosticcontext":"a5a6fd35-fc61-45b9-a3b1-4a29b9a8a1ff","time":"2019-12-09T15:37:00.2252213Z"}}}
Failed to add release annotation. Error: Failed to update Application Insights 'cp-test' Resource. Error: NotFound - Not Found (CODE: 404)
##[debug][POST]https://management.azure.com/subscriptions/.../resourceGroups/cp-visits-test/providers/Microsoft.Web/sites/cp-visits-test/publishxml?api-version=2016-08-01
##[debug]Could not parse response: {}
##[debug]Response: undefined
```

If I repeat the deployment after some time, it succeeds. Therefore, the AppInsights **is not ready after deployment**. Is there any workaround except to add a delay to my deployment pipeline?

**Comment by Dmitry-Matveev** [comment]
@debugthings , would you know how to address this issue with release annotation?

**Comment by debugthings** [comment]
Answered in #442

**Comment by M0ns1gn0r** [comment]
@debugthings The mentioned issue doesn't seem to be related to mine. In my case `addReleaseAnnotation` does work in the end, but only after a delay of 2.5 minutes.

**Comment by debugthings** [comment]
@M0ns1gn0r sorry about that I assumed this was part of the larger issue with TLS. 

**TL;DR** You're deleting and recreating the resource quick enough that it's looking up an old id in the local memory cache.

**Long Version**
When I looked at the time frame near the one in the original post I can see that there was a number of operations over the course of the day. It seems like you are creating and deleting the resource each time you deploy.

The AppInsights instances are indexed by a GUID that is randomly generated every time you create a resource; this is known as the Application Id. When a request is made to our service with the resource name it will look up that name and store the Application Id in cache mapped to the application name.

When a delete is performed it removes the resource configuration from the central storage location however it doesn't cascade to all machines to remove it from the in memory cache as the cache should timeout organically. The same is true for a create operation, it will not broadcast this to all machines to update their in memory cache.

The time between deletes and creates was short enough to hit the condition where the old Application Id value had not expired in our in memory cache. In the table below you can see that most of these requests are under 15 minutes apart. However our in memory cache timeout is 30 minutes.

Time|Operation|Result
-- | -- | --
2019-12-09T15:23:48.9616822Z | DELETE   \*/providers/\*/components/\* | 200
2019-12-09T15:36:59.9439693Z | PUT   \*/providers/\*/components/\*/annotations | 404
2019-12-10T07:17:09.2061437Z | DELETE \*/providers/\*/components/\* | 200
2019-12-10T07:27:42.6828418Z | PUT   \*/providers/\*/components/\*/annotations | 404
2019-12-10T09:30:21.2125491Z | DELETE \*/providers/\*/components/\* | 200
2019-12-10T09:45:51.4719325Z | PUT   \*/providers/\*/components/\*/annotations | 404
2019-12-10T09:47:03.3161892Z | DELETE \*/providers/\*/components/\* | 200
2019-12-10T10:00:12.2662538Z | PUT   \*/providers/\*/components/\*/annotations | 404
2019-12-10T10:02:51.6425706Z | DELETE \*/providers/\*/components/\* | 200
2019-12-10T10:16:08.9764234Z | PUT   \*/providers/\*/components/\*/annotations | 404
2019-12-10T10:18:26.6263402Z | DELETE \*/providers/\*/components/\* | 200
2019-12-10T10:30:42.8734239Z | PUT   \*/providers/\*/components/\*/annotations | 404

**Comment by M0ns1gn0r** [comment]
@debugthings Thanks for the detailed answer!

I in fact was deleting the AppInsights, although not during each deployment, but manually from the portal. I did that to ensure that creation of the infrastructure works even when it's done from scratch.
